pkg/common/commands/resetoffset/refmappers/subscription.go

```diff
@@ -0,0 +1,155 @@
+}
+
+// Create implements the ResetOffsetRefMapperFactory interface for Subscription references.  It will return
+// a new SubscriptionRefMapper instance using the specific TopicName / GroupId mappers.  It also relies on
+// the Context having injected informers (SubscriptionInformer).
+func (f *SubscriptionRefMapperFactory) Create(ctx context.Context) ResetOffsetRefMapper {
+	return NewSubscriptionRefMapper(ctx, f.TopicNameMapper, f.GroupIdMapper)
+}
+
+//
+// SubscriptionRefMapper
+//
+
+// SubscriptionTopicNameMapper defines a function signature for mapping a Subscription to a Kafka Topic name.
+type SubscriptionTopicNameMapper func(*messagingv1.Subscription) (string, error)
+
+// SubscriptionConsumerGroupIdMapper defines a function signature for mapping a Subscription to a Kafka ConsumerGroup ID.
+type SubscriptionConsumerGroupIdMapper func(*messagingv1.Subscription) (string, error)
+
+// Verify The Subscription ResetOffsetRefMapper Implements The Interface
+var _ ResetOffsetRefMapper = &SubscriptionRefMapper{}
+
+// SubscriptionRefMapper implements the ResetOffsetRefMapper for Knative Subscriptions
+type SubscriptionRefMapper struct {
+	logger             *zap.Logger
+	subscriptionLister messaginglisters.SubscriptionLister
+	topicNameMapper    SubscriptionTopicNameMapper
+	groupIdMapper      SubscriptionConsumerGroupIdMapper
+}
+
+// NewSubscriptionRefMapper returns an initialized ResetOffsetSubscriptionRefMapper
+func NewSubscriptionRefMapper(ctx context.Context,
+	topicNameMapper SubscriptionTopicNameMapper,
+	groupIdMapper SubscriptionConsumerGroupIdMapper) *SubscriptionRefMapper {
+
+	// Get The Logger From Context
+	logger := logging.FromContext(ctx).Desugar()
+
+	// Get The Subscription Informer From Context (Context Must Have Injected Informers From SharedMain())
+	subscriptionInformer := subscriptioninformers.Get(ctx)
+
+	// Return An Initialized ResetOffsetSubscriptionRefMapper
+	return &SubscriptionRefMapper{
+		logger:             logger,
+		subscriptionLister: subscriptionInformer.Lister(),
+		topicNameMapper:    topicNameMapper,
+		groupIdMapper:      groupIdMapper,
+	}
+}
+
+// MapRef implements the ResetOffsetRefMapper interface for Subscription references. It will return an
+// error in all cases other than successfully mapping the ResetOffset.Spec.Ref to a Kafka Topic / Group.
+func (m *SubscriptionRefMapper) MapRef(resetOffset *kafkav1alpha1.ResetOffset) (string, string, error) {
+
+	// Validate The ResetOffset
+	if resetOffset == nil {
+		m.logger.Warn("Received nil ResetOffset argument")
+		return "", "", fmt.Errorf("unable to map nil ResetOffset")
+	}
+
+	// Get The ResetOffset Ref From Spec & Enhance Logger
+	ref := resetOffset.Spec.Ref
+	logger := m.logger.With(zap.Any("Ref", ref))
+
+	// Validate The Reference
+	if !strings.HasPrefix(ref.APIVersion, messaging.GroupName) ||
+		ref.Kind != "Subscription" ||
+		len(ref.Name) <= 0 {
+		m.logger.Warn("Received ResetOffset with invalid Subscription reference")
```

Comment:
What about changing this to "ResetOffset supports only subscriptions"

Reply:
I was using the "invalid Subscription" language to also cover the case where it IS a subscription but the Name was missing.  I will break it up though into two different checks though with better/detailed logging/errors in each case - because as you indicate it would be good to make it clear that it's only expecting Subscriptions.